### PR TITLE
refactor resultset reply

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

File: tests/flow/test_bolt.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 from neo4j import GraphDatabase
 from neo4j.spatial import WGS84Point
 import neo4j.graph
-from neo4j.debug import watch
+# from neo4j.debug import watch
```
Comment (swilly22) [inline tests/flow/test_bolt.py:5]: Remove ?

File: tests/flow/test_bolt.py [hunk 2]
```diff
@@ -12,7 +12,7 @@ def __init__(self):
         port = 7687 + 6379 - self.env.port
         global bolt_con
         bolt_con = GraphDatabase.driver(f"bolt://localhost:{port}", auth=("falkordb", ""))
-        self.watcher = watch("neo4j")
+        # self.watcher = watch("neo4j")
```
Comment (swilly22) [inline tests/flow/test_bolt.py:15]: Remove ?

File: src/resultset/resultset.h [hunk 3]
```diff
@@ -16,7 +16,7 @@
 #define RESULTSET_OK 1
 #define RESULTSET_FULL 0
 
-typedef struct {
+typedef struct ResultSet {
```
Comment (swilly22) [inline src/resultset/resultset.h:19]: I'm not sure we need to include the name of the struct at the beginning and at the end, one of them should be removed.

File: src/resultset/formatters/resultset_replyverbose.h [hunk 4]
```diff
@@ -9,18 +9,16 @@
...
 void ResultSet_ReplyWithVerboseHeader
 (
-	RedisModuleCtx *ctx,
-	bolt_client_t *bolt_client,
-	const char **columns,
-	uint *col_rec_map
+	ResultSet *set
 );
 
 void ResultSet_EmitVerboseRow
```
Comment (swilly22) [inline src/resultset/formatters/resultset_replyverbose.h:15]: I'm missing documentation for each function in this header file

File: src/resultset/formatters/resultset_replyverbose.c [hunk 5]
```diff
@@ -217,33 +218,43 @@ static void _ResultSet_VerboseReplyWithVector
...
-	GraphContext *gc,
-	SIValue **row,
-	uint numcols
+	ResultSet *set,
+	SIValue **row
 ) {
+	RedisModuleCtx *ctx = set->ctx;
 	// Prepare return array sized to the number of RETURN entities
-	RedisModule_ReplyWithArray(ctx, numcols);
+	RedisModule_ReplyWithArray(ctx, set->column_count);
```
Comment (swilly22) [inline src/resultset/formatters/resultset_replyverbose.c:226]: if possible access `ResultSet` fields via functions, letting a reply formatter tinker with the `ResultSet` internals isn't recomendded

File: src/resultset/formatters/resultset_replyverbose.c [hunk 6]
```diff
@@ -217,33 +218,43 @@ static void _ResultSet_VerboseReplyWithVector
...
-		RedisModule_ReplyWithStringBuffer(ctx, columns[i], strlen(columns[i]));
+		RedisModule_ReplyWithStringBuffer(ctx, set->columns[i], strlen(set->columns[i]));
 	}
 }
+
+void ResultSet_EmitVerboseStats
+(
+	ResultSet *set
+) {
+	ResultSetStat_emit(set->ctx, &set->stats);
```
Comment (swilly22) [inline src/resultset/formatters/resultset_replyverbose.c:259]: This is strange, to have a formatter call the result-set in order to emit data,
I understand this was done probably for code reuse, if that's indeed the case we need to find a better place for `ResultSetStat_emit` logic to reside at.

File: src/resultset/formatters/resultset_replynop.h [hunk 7]
```diff
@@ -9,18 +9,16 @@
...
-    RedisModuleCtx *ctx,
-    bolt_client_t *bolt_client,
-    GraphContext *gc,
-    SIValue **row,
-    uint numcols
+    ResultSet *set,
+    SIValue **row
 );
 
+void ResultSet_EmitNOPStats
```
Comment (swilly22) [inline src/resultset/formatters/resultset_replynop.h:21]: As with the other formatters header files, please make sure each function in the header file is documented.

File: src/resultset/formatters/resultset_formatter.h [hunk 8]
```diff
@@ -33,16 +32,19 @@ typedef enum {
 	VALUE_POINT = 11,
 } ValueType;
 
+typedef struct ResultSet ResultSet;
```
Comment (swilly22) [inline src/resultset/formatters/resultset_formatter.h:35]: is this to avoid cyclic inclusion ?

**Review by swilly22 — CHANGES_REQUESTED** [review]
Nicely done, please have a look at the few comments I have.

File: tests/flow/test_bolt.py (hunk 1 above)
Comment (AviAvni) [reply tests/flow/test_bolt.py:5]: no I need this when debug

File: tests/flow/test_bolt.py (hunk 2 above)
Comment (AviAvni) [reply tests/flow/test_bolt.py:15]: no

File: src/resultset/resultset.h (hunk 3 above)
Comment (AviAvni) [reply src/resultset/resultset.h:19]: need otherwise we can't declare typedef later

**Pull request merged**